### ISSUE TITLE
cloud_provider credentials exposed to lowest privilege users

### ISSUE THREAD

**Issue by verikca** [rancher/rancher#19653]
cloud_provider credentials (in the "edit cluster") exposed in clear text to any user with the lowest privileges to the system. In the LDAP case this literally means everyone.

*Ways to reproduce:*
1. Install Rancher cluster HA
2. Create a subordinate cluster with either Openstack driver or Custom cluster
2a. Set up the cloud_provider under "Edit cluster" either at the time of or after creating the cluster, provide cloud credentials using YAML there (primarily for attached storage purpose).
3. Create a user with global privileges "Login Access" only
4. Grant the user privileges "View Workloads" only to Default project of subordinate cluster.
5. Log in as the user above
6. Go to the Clusters page, click tree ellipse on the subordinate cluster and choose "View in API".
6a. Alternatively Go to the Clusters page, click Edit for subordinate cluster and choose "Edit as YAML"
7. Behold the credentials to the cloud in clear text.

*Expected behavior:*
Only cluster owner or similar role should be able to see and edit the "Cluster Options" or see appliedSpec.rancherKubernetesEngineConfig block of cluster parameters in API.

Imported clusters work perfectly well without exposing this block to any Rancher user.

Alternatively the cloud provider "password" parameter needs to be redacted for output but it becomes complicated as every cloud provider has it in the different place.

P.S. In every cluster including imported ones these credentials can be found under kube-system/cluster-state or kube-system/full-cluster-state config maps. So it makes sense to limit the scope of users to those who can read the config maps from the kube-system project.

---
| Useful | Info |
| :-- | :-- |
|Versions|Rancher `v2.2.1` UI: `v2.2.19` |
|Route|`global-admin.clusters.index`|

Tested with Rancher v2.2.1 and v2.1.6, with Openstack cloud_provider with Openstack driver and Custom cluster as well as VmWare cloud_provider with Custom cluster.

**Comment by alena1108** [comment]
@verikca by adding a user to a project, it gets permissions to view clusters, no matter how limited his permission for the projects are (View Workloads in your case). But we should definitely address hiding sensitive fields like password from the API; it should be done regardless of who the viewer is. We have a way to enforce this behavior in API by adding a special tag to a field definition (https://github.com/rancher/types/blob/master/apis/management.cattle.io/v3/cluster_types.go#L98)

We should add the same tag for the Cloud providers defined in types (like openstack):
https://github.com/rancher/types/blob/master/apis/management.cattle.io/v3/rke_types.go#L489

For the custom cloud providers, it would require a custom logic, like passing an extra field to the cloud config, where all the sensitive fields are mentioned. So our API knows how to hide it.

**Comment by alena1108** [comment]
@mrajashree in addition to types fixes, we should also put a rancher fix similar to pr linked to https://github.com/rancher/rancher/pull/18973. When replace=true is sent on the update, we don't want the hidden password fields to be reset.

**Comment by mrajashree** [comment]
@alena1108 As per offline discussion, addressing the password getting overriden on replace=true with the third approach you suggested for this issue https://github.com/rancher/rancher/issues/18964#issuecomment-473499707

**Comment by alena1108** [comment]
Can be tested on latest master.

**Comment by sangeethah** [comment]
Able to reproduce the issue when testing with v2.2.2
1. Create a Azure cluster with cloud provider.
2. Once the cluster gets to "active" state , edit cluster and use "Edit as YAML" option to see the cluster yml. "aad_client_secret: " under "azureCloudProvider" section is available in clear text.
3. Use "View in Api" option .  "aad_client_secret: " (part of "azureCloudProvider") is available in clear text.

This issue is fixed when  testing with v2.2.3-rc7.
1. Create a Azure cluster with cloud provider.
2. Once the cluster gets to "active" state , edit cluster and use "Edit as YAML" option to see the cluster yml. "aad_client_secret: " is not present anymore in  "azureCloudProvider".
3. Use "View in Api" option .  "aad_client_secret:" is not available as part of "azureCloudProvider" anymore.
4. on edit page, user is not presented with the "aadClientSecret" parameter anymore. 
Edit cluster ( In my case I disabled recurring snapshots)  and "Save". Following error message is presented to the user

<img width="1361" alt="Screen Shot 2019-05-07 at 7 31 28 PM" src="https://user-images.githubusercontent.com/4266958/57345243-f7a06980-70fe-11e9-8699-f3608109df09.png">

**Comment by soumyalj** [comment]
Tested with master. 
Created Azure cluster with cloud provider set. 
After the cluster was active, viewed the yaml file of the cluster. It was observed that `aad_client_secret` was empty as below:

```
azureCloudProvider: 
    aad_client_id: "abcd"
    cloud_provider_backoff: false
    cloud_provider_backoff_duration: 0
    cloud_provider_backoff_exponent: 0
    cloud_provider_backoff_jitter: 0
    cloud_provider_backoff_retries: 0
    cloud_provider_rate_limit: false
    cloud_provider_rate_limit_bucket: 0
    cloud_provider_rate_limit_qps: 0
    maximum_load_balancer_rule_count: 0
    subscription_id: "abcd"
    tenant_id: "efgh"
    use_instance_metadata: false
    use_managed_identity_extension: false
    aad_client_cert_password: ""
    aad_client_secret: ""

```
In the API , aad_client_secret is not present as part of  "azureCloudProvider" spec:

```
"cloudProvider": {
"azureCloudProvider": {
"aadClientCertPath": "",
"aadClientId": "abcd",
"cloud": "",
"cloudProviderBackoff": false,
"cloudProviderBackoffDuration": 0,
"cloudProviderBackoffExponent": 0,
"cloudProviderBackoffJitter": 0,
"cloudProviderBackoffRetries": 0,
"cloudProviderRateLimit": false,
"cloudProviderRateLimitBucket": 0,
"cloudProviderRateLimitQPS": 0,
"location": "",
"maximumLoadBalancerRuleCount": 0,
"primaryAvailabilitySetName": "",
"primaryScaleSetName": "",
"resourceGroup": "",
"routeTableName": "",
"securityGroupName": "",
"subnetName": "",
"subscriptionId": "efgh",
"tenantId": "ijkl",
"type": "/v3/schemas/azureCloudProvider",
"useInstanceMetadata": false,
"useManagedIdentityExtension": false,
"vmType": "",
"vnetName": "",
"vnetResourceGroup": ""
},

```
Also, the user can edit the cluster by changing any cluster settings without providing the `aadClientSecret` value again .
